File: new_dir/templates/C/skirl.cpp
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#include "skirl.h"

Quad** readArray(const char* path, int* w, int* h)
{
    FILE* F = fopen(path, "rb");
    if(!F)
        return NULL;
    // Первый проход по файлу - определение размера массива
    struct
    {
        int flags;
        int size;
        short int angle;
        short int reserve;
        int time;
    } header;
    int nLines = 0;
    while(!feof(F))
    {
        int res = fread(&header, sizeof(header), 1, F);
        if(res == 0)
            break;
        nLines += 1;
        fseek(F, 4*header.size, SEEK_CUR);  // Переход к след. строке
    }
    Quad** out = (Quad**)malloc(nLines*sizeof(Quad*));
    // Второй проход - чтение данных
    fseek(F, 0, SEEK_SET);
    int y=0;
    while(!feof(F))
    {
        int res = fread(&header, sizeof(header), 1, F);
        if(res == 0)
            break;
        out[y] = (Quad*)malloc(header.size*sizeof(Quad));
        for(int a=0; a<header.size; a++)
        {
            fread(&out[y][a], sizeof(Quad), 1, F);
        }
        y+=1;
    }
    fclose(F);
    *w=header.size;
    *h=nLines;
    return out;
}
//----------------------------------------------------------------
int writeArray(Quad** ar, const char* path, int w, int h)
{
    FILE* F=fopen(path, "wb");
    if(!F)
        return -1;
    struct
    {
        int flags;
        int size;
        short int angle;
        short int reserve;
        int time;
    } header;
    for(int y=0; y<h; y++)
    {
        header.flags = 0;
        header.size = w;
        header.angle=floor(y*10000.0/2400.0);
        header.reserve = 0xDEAD;
        header.time = y;
        fwrite(&header, sizeof(header), 1, F);
        for(int x=0; x<w; x++)
        {
            fwrite(&ar[y][x], sizeof(Quad), 1, F);
        }
    }
    fclose(F);
    return 0;
}
//----------------------------------------------------------------
// Освобождает память, выделенную для массива.
// Параметры:
// ar — указатель на массив
// h — количество строк в массиве
void clearArray(Quad** ar, int h)
{
    for(int a=0; a<h; a++)
        free(ar[a]);
    free(ar);
}

File: new_dir/templates/python/skirl.py
# coding=utf-8

from struct import pack,unpack
from math import sqrt,floor

# Параметры:
#  path — путь к файлу
# Возвращаемое значение: двумерный массив отсчетов, оформленный как
# массив из массивов.
def readArray(path):
    res = []
    try:
        F=open(path, "rb")
    except IOError:
        return
    while True:
        s = F.read(16)
        if(len(s) < 16):
            break
        header = unpack("iihhi", s)
        flags = header[0]
        size = header[1]
        angle = header[2]
        time = header[4]
        line=[]
        for a in range(size):
            s = F.read(4);
            if(len(s) < 4):
                break
            pt = unpack("hh", s)
            line.append(list(pt));
        res.append(line)
    F.close()
    return res

# Параметры:
#  field — массив, записываемый в файл
#  path — путь к файлу, в который следует сохранить данные
def writeArray(field, path):
    F=open(path, "wb")
    for a in range(len(field)):
        line = field[a]
        size=len(line)
        angle = floor(a*10000.0/2400.0)
        time=100
        s=pack("iihHi", 0, size, angle, 0xDEAD, time)
        F.write(s)
        for b in range(len(line)):
            s = pack("hh", line[b][0], line[b][1])
            F.write(s)
    F.close()
